File: src/main/java/io/lsdconsulting/lsd/distributed/interceptor/captor/messaging/MessagingHeaderRetriever.java
package io.lsdconsulting.lsd.distributed.interceptor.captor.messaging;

import io.lsdconsulting.lsd.distributed.interceptor.captor.header.Obfuscator;
import lombok.RequiredArgsConstructor;
import org.springframework.messaging.Message;

import java.util.Collection;
import java.util.List;
import java.util.Map;

import static java.util.Collections.emptyList;
import static java.util.stream.Collectors.toMap;

@RequiredArgsConstructor
public class MessagingHeaderRetriever {

    private final Obfuscator obfuscator;

    public Map<String, Collection<String>> retrieve(final Message<?> message) {
        return obfuscator.obfuscate(message.getHeaders().entrySet().stream()
                .collect(toMap(Map.Entry::getKey, e -> e.getValue() != null ? List.of(e.getValue().toString()) : emptyList())));
    }
}

File: src/main/java/io/lsdconsulting/lsd/distributed/interceptor/captor/messaging/MessagingHeaderRetriever.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor.messaging

import io.lsdconsulting.lsd.distributed.interceptor.captor.header.Obfuscator
import lombok.RequiredArgsConstructor
import org.springframework.messaging.Message

@RequiredArgsConstructor
class MessagingHeaderRetriever(
    private val obfuscator: Obfuscator
) {
    fun retrieve(message: Message<*>): Map<String, Collection<String>> =
        obfuscator.obfuscate(message.headers.entries.associate {
            it.key to (it.value?.let { _ -> listOf(it.value.toString()) } ?: emptyList())
        })
}
